server/fleet/app.go

```diff
@@ -239,6 +258,29 @@ func (c *AppConfig) ApplyDefaults() {
 	c.WebhookSettings.Interval.Duration = 24 * time.Hour
 }
 
+// EnableStrictDecoding enables strict decoding of the AppConfig struct.
+func (c *AppConfig) EnableStrictDecoding() { c.strictDecoding = true }
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (c *AppConfig) UnmarshalJSON(b []byte) error {
+	// Decode any legacy settings that might be present.
+	var legacy legacyConfig
+	if err := json.Unmarshal(b, &legacy); err != nil {
+		return err
+	}
+
+	// TODO(roperzh): define and assign legacy settings to new fields. E.g:
+	// c.Features = legacy.HostSettings
+
+	// Decode the config, overriding any legacy settings.
+	type alias *AppConfig
```

Comment:
I understand the need for a distinct type so that it doesn't call `UnmarshalJSON` in a loop, but maybe a clearer name would help someone confused by that alias type (appCfgNoCustomUnmarshal? and/or an explanation comment, this is a bit of a non-obvious, tricky thing worth explaining ).

Reply:
you're right, and on top of that it's not a type alias it's a type definition. Updated.